### PR TITLE
pmap: make CachingPool the default for anonymous functions

### LINKED ISSUE #21946: Proposal: Make creation of CachingPool a default for pmap
Ref #21940

In most parallel operations I've come across, there is at least some data that moves to the processor. Because it appears that `pmap` sends function arguments to workers for every block (determined by `block_size`, default=1), this means that `pmap`s default behavior results in large inefficiencies when functions within the `pmap` block reference large data structures.

@amitmurthy has suggested, and testing has confirmed, that the use of `Distributed.CachingPool` will reduce the amount of data transferred to functions called within a `pmap` block.

The use of caching pools incurs a small amount of overhead, which is more apparent when `pmap` doesn't result in data transfers of appreciable size. This is less common in real-world situations, though.

This proposal is to modify `pmap` so that it creates a caching pool equal to the number of workers by default, with an optional override when `pmap` is used without data transfer. This provides the expected performance gains from `pmap` use under most real-world conditions out-of-the-box, with the flexibility to tune things in these less-frequent cases.

Right now, there is [a proposal](https://github.com/JuliaLang/julia/issues/19578) to deprecate `@parallel` for loops. `@parallel` does not suffer from this issue, and therefore provides better performance than `pmap` when data is processed by workers. This proposal would insure that `pmap` is at least as performant as `@parallel` once the latter is deprecated.

### PR DESCRIPTION
**Pull request by amitmurthy**
Closes https://github.com/JuliaLang/julia/issues/21946 .

This is useful when the mapping function captures a large amount of data. Automatically using a `CachingPool` in such circumstances avoids sending said data over the network for each call.

### CODE REVIEW

File: base/distributed/pmap.jl [hunk 1]
```diff
@@ -293,3 +306,18 @@ function batchsplit(c; min_batch_count=1, max_batch_size=100)
...
+        name = tn.mt.name
+        mod = tn.module
+        if t.super === Function &&
+            unsafe_load(Base.unsafe_convert(Ptr{UInt8}, tn.name)) == UInt8('#') &&
+            (!isdefined(mod, name) || t != typeof(getfield(mod, name)))
+            return true
+        end
+    end
+    return false
+end
```
Comment (amitmurthy) [inline base/distributed/pmap.jl]: @vtjnash / @JeffBezanson , this is more or less an adaptation of the code in https://github.com/JuliaLang/julia/blob/bf47fd90f50bfda87fad5a501829e91a60c7d0d0/base/serialize.jl#L500-L517 without a deeper understanding of how this is supposed to work. Can you review this for correctness? Or suggest the correct way to test if a function is an anonymous function?

**Comment by vtjnash** [comment]
Why did we stop on this?

**Comment by StefanKarpinski** [comment]
Bump. @amitmurthy: what's needed to make progress on this?

**Comment by vtjnash** [comment]
AFAIK, most of this was included in https://github.com/JuliaLang/julia/pull/33892, but now Distributed is not in Base anymore

**Pull request closed without merging**